File: mobile/gateway/test/runtime-status.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { projectRuntimeStatus, mobileRuntimeStatus } from '../src/runtime-status.mjs';

test('runtime status projects only the bounded mobile contract', () => {
  const status = projectRuntimeStatus({
    phase: 'ready', connection: 'connected', lastConfirmedAt: '2026-03-20T10:00:00.000Z', errorCode: 'NONE', updatePending: true,
    instance: { instanceId: 'runtime-1', namespaceId: 'desktop', channel: 'stable', protocolVersion: 1, runtimeRevision: 'a'.repeat(64), startedAt: '2026-03-20T09:00:00.000Z', rawPath: 'C:/private' },
    installation: { kind: 'installed', label: 'Zyra Desktop', appVersion: '1.2.3', token: 'secret' }, sessions: [{ id: 'private' }]
  });
  assert.deepEqual(status, {
    phase: 'ready', connection: 'connected', lastConfirmedAt: '2026-03-20T10:00:00.000Z', errorCode: 'NONE', updatePending: true,
    instance: { instanceId: 'runtime-1', namespaceId: 'desktop', channel: 'stable', protocolVersion: 1, runtimeRevision: 'a'.repeat(64), startedAt: '2026-03-20T09:00:00.000Z' },
    installation: { kind: 'installed', label: 'Zyra Desktop', appVersion: '1.2.3' }
  });
  assert.equal(projectRuntimeStatus({ phase: 'ready' }), undefined, 'missing connection is unknown, never ready');
  assert.equal(projectRuntimeStatus(undefined), undefined);
});

test('phone connection health wins over a cached desktop status', () => {
  const configured = { phase: 'ready', connection: 'connected', installation: { kind: 'development', label: 'Development-test' } };
  const value = mobileRuntimeStatus({ phase: 'failed', connection: 'disconnected' }, configured);
  assert.equal(value.connection, 'disconnected');
  assert.equal(value.phase, 'failed');
  assert.equal(value.installation.label, 'Development-test');
  assert.equal(mobileRuntimeStatus(undefined, undefined), undefined);
  assert.equal(projectRuntimeStatus({ phase: 'idle', connection: 'unknown', installation: configured.installation }).connection, 'unknown');
});

File: mobile/android/app/src/test/java/dev/zyra/mobile/RuntimeStatusTest.kt
package dev.zyra.mobile

import dev.zyra.mobile.data.RuntimeConnection
import dev.zyra.mobile.data.RuntimePhase
import dev.zyra.mobile.data.RuntimeStatus
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class RuntimeStatusTest {
    @Test fun `missing hello runtime status stays unknown`() {
        val status = RuntimeStatus.fromHello(JSONObject())
        assertFalse(status.known)
        assertEquals(RuntimePhase.Unknown, status.phase)
    }

    @Test fun `phone freshness does not depend on matching the PC clock`() {
        val payload = JSONObject("""{"phase":"ready","connection":"connected","lastConfirmedAt":"2026-01-01T00:00:00Z"}""")
        val serverTime = java.time.Instant.parse("2026-01-01T00:00:02Z").toEpochMilli()
        val localTime = serverTime + 3600000
        val status = RuntimeStatus.from(payload, serverTime, localTime)
        assertTrue(status.isLive(localTime))
        assertEquals("Live", status.syncLabel(localTime))
        assertFalse(status.isLive(localTime + 46000))
        assertEquals("Status stale", status.syncLabel(localTime + 46000))
        assertEquals("Disconnected", status.copy(connection = RuntimeConnection.Disconnected).syncLabel(localTime))
    }

    @Test fun `runtime event accepts bounded public instance metadata`() {
        val status = RuntimeStatus.fromEvent(JSONObject("""{"phase":"ready","connection":"connected","updatePending":true,"instance":{"instanceId":"runtime-7","channel":"stable"},"installation":{"kind":"installed","label":"Zyra","appVersion":"1.2.3"}}"""))
        assertTrue(status.known)
        assertEquals(RuntimePhase.Ready, status.phase)
        assertEquals(RuntimeConnection.Connected, status.connection)
        assertEquals("runtime-7", status.instance?.instanceId)
        assertEquals("Zyra", status.installation?.label)
    }
}
